homeassistant/components/sensor/__init__.py

```diff
@@ -260,7 +261,17 @@ def unit_of_measurement(self) -> str | None:
 
         native_unit_of_measurement = self.native_unit_of_measurement
 
-        if native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+        # Only convert Kelvin if the sensor is for temperature.
+        # Color temperature is measured in Kelvin, but shouldn't be a 'temperature' class.
+        is_temperature = (
+            native_unit_of_measurement in TEMPERATURE_UNITS
+            and self.device_class == DEVICE_CLASS_TEMPERATURE
+        ) or native_unit_of_measurement in (
+            TEMP_CELSIUS,
+            TEMP_FAHRENHEIT,
+        )
```

Comment:
This form might make it clearer:
```python
if (
    native_unit_of_measurement in (TEMPERATURE_UNITS)
    and (self.device_class == DEVICE_CLASS_TEMPERATURE or native_unit_of_measurement != TEMP_KELVIN)
):
```

Reply:
I'm not sure if the explicit exclusion make it much clearer, to be honest, but I'm happy to give it a try later (there's some ordering issue in conditions because otherwise we end up in an infinite recursion with some sensors defining their device class based on their UoM.)

My main thought behind this block is that after the deprecation is removed, the whole `or` part is gone — well-behaved sensors only enters the first part of the condition.